Userland/Applications/FTPServer/FTPServerClient.cpp

```diff
@@ -0,0 +1,1169 @@
+    }
+
+    String file = String::join(" ", params);
+
+    StringBuilder builder;
+    builder.append(m_working_dir);
+    builder.append("/");
+    builder.append(file);
+    auto path = LexicalPath::canonicalized_path(builder.to_string());
+
+    auto result = Core::File::remove(path, Core::File::RecursionMode::Disallowed, false);
+    if (result.is_error()) {
+        send_file_action_not_taken();
+        return;
+    }
+
+    send_file_action_ok();
+}
+
+void FTPServerClient::handle_cwd_command(Vector<String> params)
+{
+    if (params.size() < 1) {
+        send_command_not_implemented_for_parameter();
+        return;
+    }
+
+    if (!params.at(0).starts_with("/")) {
+        StringBuilder builder;
+        builder.append(m_working_dir);
+        builder.append("/");
+        builder.append(params.at(0));
+        (void)params.take_first();
+
+        params.prepend(builder.to_string());
+    }
+
+    String path = String::join(" ", params);
+
+    m_working_dir = LexicalPath::canonicalized_path(path);
+    send_file_action_ok();
+}
+
+void FTPServerClient::handle_cdup_command()
+{
+    StringBuilder builder;
+    builder.append(m_working_dir);
+    builder.append("/..");
+    m_working_dir = LexicalPath::canonicalized_path(builder.to_string());
+    send_file_action_ok();
+}
+
+void FTPServerClient::handle_list_command(Vector<String> params)
+{
+    if (params.size() < 1) {
+        send_directory_content(m_working_dir, false);
+        return;
+    }
+
+    String path = String::join(" ", params);
+
+    if (path.starts_with('-')) {
+        send_directory_content(m_working_dir, false);
+        return;
+    }
+
+    send_directory_content(path, false);
+}
+
+void FTPServerClient::handle_nlst_command(Vector<String> params)
+{
+    if (params.size() < 1) {
+        send_directory_content(m_working_dir, true);
+        return;
+    }
+
+    String path = String::join(" ", params);
+
+    if (path.starts_with('-')) {
+        send_directory_content(m_working_dir, true);
+        return;
+    }
+
+    send_directory_content(path, true);
+}
+
+void FTPServerClient::handle_type_command(Vector<String> params)
+{
+    if (params.size() < 1) {
+        send_invalid_parameters();
+        return;
+    }
+
+    String type = params.at(0);
+
+    if (!type.equals_ignoring_case("I") && !type.equals_ignoring_case("A")) {
+        send_command_not_implemented_for_parameter();
+        return;
+    }
+
+    m_transfer_type = type;
+    send_ok();
+}
+
+void FTPServerClient::handle_stru_command(Vector<String> params)
+{
+    if (params.size() < 1) {
+        send_invalid_parameters();
+        return;
+    }
+
+    String type = params.at(0);
+
+    if (!type.equals_ignoring_case("F")) {
+        send_command_not_implemented_for_parameter();
+        return;
+    }
+
+    m_file_structure = type;
+    send_ok();
+}
+
+void FTPServerClient::handle_pasv_command()
+{
+    auto address = IPv4Address::from_string(get_current_ip()).value();
```

Comment:
`get_current_ip()` can fail and return an empty string. Will this fail or fall back to all interfaces `0.0.0.0` ?

Reply:
Swapped to using `m_socket->address()`, so this method is removed.